pkg/controller/namespace/deletion/namespaced_resources_deleter.go

```diff
@@ -25,66 +25,190 @@ import (
 	"k8s.io/kubernetes/pkg/api/errors"
 	"k8s.io/kubernetes/pkg/api/unversioned"
 	"k8s.io/kubernetes/pkg/api/v1"
-	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/core/internalversion"
 	"k8s.io/kubernetes/pkg/client/typed/dynamic"
 	"k8s.io/kubernetes/pkg/runtime"
 	"k8s.io/kubernetes/pkg/util/sets"
 
 	"github.com/golang/glog"
 )
 
-// contentRemainingError is used to inform the caller that content is not fully removed from the namespace
-type contentRemainingError struct {
+func NewNamespacedResourcesDeleter(nsClient internalversion.NamespaceInterface,
+	clientPool dynamic.ClientPool, opCache OperationNotSupportedCache,
+	gvrs []unversioned.GroupVersionResource, finalizerToken api.FinalizerName,
+	deleteNamespaceWhenDone bool) *NamespacedResourcesDeleter {
+	return &NamespacedResourcesDeleter{
+		nsClient:                nsClient,
+		clientPool:              clientPool,
+		opCache:                 opCache,
+		groupVersionResources:   gvrs,
+		finalizerToken:          finalizerToken,
+		deleteNamespaceWhenDone: deleteNamespaceWhenDone,
+	}
+}
+
+// NamespacedResourcesDeleter is used to delete all resources in a given namespace.
+type NamespacedResourcesDeleter struct {
+	// Client to manipulate the namespace.
+	nsClient internalversion.NamespaceInterface
+	// Dynamic client to list and delete all namespaced resources.
+	clientPool dynamic.ClientPool
+	// Cache of what operations are not supported on each group version resource.
+	opCache OperationNotSupportedCache
+	// Group version resources that should be operated on.
+	groupVersionResources []unversioned.GroupVersionResource
+	// The finalizer token that should be removed from the namespace
+	// when all resources in that namespace have been deleted.
+	finalizerToken api.FinalizerName
+	// Also delete the namespace when all resources in the namespace have been deleted.
+	deleteNamespaceWhenDone bool
```

Comment:
In federation, we can not delete the namespace when all resources in that namespace are deleted.
We also have to first delete the namespace from all underlying clusters and then delete the namespace. Hence, we set this to true in federation.